feat: 4.2 article (#954)
withastro/astro.build: src/content/blog/astro-420.mdx

---
title: "Astro 4.2"
description: "Astro 4.2 is out now! This release includes two new experimental features to try out, improvements to accessibility rules, and more."
publishDate: "January 18, 2024"
authors:
  - erika
  - ema
  - matthew
  - nate
  - bjorn
coverImage: "/src/content/blog/_images/astro-420/post-header-4.2.webp"
socialImage: "/src/content/blog/_images/astro-420/og-image-4.2.webp"
lang: "en"
---

> Astro 4.2 is now available! This release includes new experimental features to try out, improvements to accessibility rules, the ability for remark plugins to customize image optimization in Markdown and many more improvements and bug fixes.

Reviewer: Missing Oxford comma
```suggestion
Astro 4.2 is now available! This release includes new experimental features to try out, improvements to accessibility rules, the ability for remark plugins to customize image optimization in Markdown, and many more improvements and bug fixes.
```

## How to upgrade

To take advantage of the latest features, make sure you're running the latest version of Astro. You can upgrade to Astro 4.2 by running the `@astrojs/upgrade` command:

```sh
npx @astrojs/upgrade
```

or by running the upgrade command for your package manager:

```sh
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## (Experimental) Prerendering pages using the Speculation Rules API

Thanks to [Ross Robino](https://github.com/rossrobino), Astro's prefetch feature now has experimental support for prerendering pages using the currently Chromium-exclusive [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API). This API allows you to prerender and even run client-side JavaScript on pages that the user is likely to visit next, allowing for a faster browsing experience.

To enable this feature, add the following to your `astro.config.mjs` file:

```diff
import { defineConfig } from 'astro/config';

// https://astro.build/config
export default defineConfig({
+    prefetch: true,
+    experimental: {
+        clientPrerender: true,
+    },
});
```

> Then, in script tags, you can use the `document.prerendering` property to check if the page is being prerendered:

Reviewer 2: This is what he told me  :sweat_smile: 

Tech review would be helpful! (And, if necessary, rework the config reference docs)

```html
<script>
		if (document.prerendering) {
				// prerendering is enabled
		}
</script>
```

> 

Author: Is there a missing word? "For more *information* about ..."?

> 

Reviewer 2: Either works! Both are correct, so go with whatever your prefer

As this is a breaking change, we've added a new `experimental` option to the `astro.config.mjs` file to opt-in to the new behavior:

```diff
import { defineConfig } from 'astro/config';

// https://astro.build/config
export default defineConfig({
+    experimental: {
+        globalRoutePriority: true,
+    },
});
```

## Improved accessibility rules

The accessibility rules added to the Astro Dev Toolbar in 4.0 have been improved to be more accurate. Thanks to [Oliver Speir](https://github.com/OliverSpeir) for meticulously reading through the WCAG guidelines and improving these rules! If you had any accessibility issues wrongfully reported by the Dev Toolbar in previous releases, you may want to check again to see if they've been fixed after this update.

## Customizable image optimization in Markdown

> Also by Oliver, remark plugins can now customize how images are optimized in Markdown files. Previously, every image included in Markdown files using the native syntax (`![alt](src)`) would use the default settings for image optimization. After this update, remark plugins can now add properties to `image` nodes to customize how they're optimized.

Reviewer 2: ```suggestion
Also by Oliver, remark plugins can now customize how images are optimized in Markdown files. Previously, every image included in Markdown files using the native syntax (`![alt](src)`) used Astro's default settings for image optimization. With this update, remark plugins can now add properties to `image` nodes to customize how they are optimized.
```

For example, the following remark plugin will set the `width` and `height` properties of every image to `100`:

```js
import { visit } from "unist-util-visit";

export default function remarkPlugin() {
	return (tree) => {
		visit(tree, 'image', (node) => {
			node.data.hProperties = node.data.hProperties || {};
			node.data.hProperties.width = "100";
			node.data.hProperties.height = "100";
		});
	};
}
```

This is currently exclusive to Markdown files, but we're hoping to add support for MDX as well in a future release.

## Bug Fixes

> As always, additional bug fixes are included in this release. Check out the [release notes](https://github.com/withastro/astro/blob/refs/heads/main/packages/astro/CHANGELOG.md#TODO) to learn more.

Author: Yep, will do once the release is out!

Thanks to [Ross Robino](https://github.com/rossrobino), Astro's `prefetch` feature now has experimental support for prerendering pages using the currently Chromium-exclusive [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API). This API allows you to prerender pages on the client, and even run client-side JavaScript on pages that the user is likely to visit next, allowing for a faster browsing experience.

To enable this feature, add the following to your `astro.config.mjs` file:

```diff
import { defineConfig } from 'astro/config';

// https://astro.build/config
export default defineConfig({
+    prefetch: true,
+    experimental: {
+        clientPrerender: true,
+    },
});
```

This will respect your existing `prefetch` configuration options, but will now prerender all links with the `data-astro-prefetch` attribute on the client, instead. A built-in fallback is provided for browsers that do not yet support the Speculation Rules API.

Additionally, you can use the `document.prerendering` property in `<script>` tags to check whether the page is being prerendered:

```html

> <script>
> 		if (document.prerendering) {
> 				// prerendering is enabled
> 		}
> </script>

Reviewer 2: ```suggestion
<script>
  if (document.prerendering) {
    // prerendering is enabled
  }
</script>
```
Please adjust this however you prefer, but it just seemed like all those tabs were *really* spacious. Can probably get away with fewer tabs, or those as spaces instead. Do whatever looks right!

![image](https://github.com/withastro/astro.build/assets/5098874/d0296dbd-9189-494d-b9bf-5a1b02afd32e)

publishDate: "January 18, 2024"
authors:
  - erika
  - ema
  - matthew
  - nate
  - bjorn
coverImage: "/src/content/blog/_images/astro-420/post-header-4.2.webp"
socialImage: "/src/content/blog/_images/astro-420/og-image-4.2.webp"
lang: "en"
---

Astro 4.2 is now available! This release includes new experimental features to try out, improvements to accessibility rules, the ability for remark plugins to customize image optimization in Markdown, and many more improvements and bug fixes.

Highlights include:

- **[(Experimental) Support for prerendering pages using the Speculation Rules API](#experimental-prerendering-pages-using-the-speculation-rules-api)**
- **[(Experimental) Reworked routing priority for injected routes](#experimental-reworked-routing-priority-for-injected-routes)**
- **[New `redirectToDefaultLocale` config option](#new-redirecttodefaultlocale-config-option)**
- **[Improved accessibility rules](#improved-accessibility-rules)**
- **[Customizable image optimization in Markdown](#customizable-image-optimization-in-markdown)**

> As a side note, this is the first release of Astro to majorly consist of outside contributions, with all but one highlights of this release being from contributors. This is a huge milestone for us, as it means that Astro is now big enough to have a thriving community of contributors capable of spearheading releases. Thank you to everyone who contributed to this release!

Reviewer 2: ```suggestion
This marks the first Astro release of almost all community-built features, with only one highlight brought to you by an Astro core maintainer. This is a huge milestone for us, as it means that Astro is now big enough to have a thriving community of contributors capable of spearheading releases. Thank you to everyone who contributed to this release!
```
There were an awful lot of "c" words between community and contributor/ing, consisting, coming from... so I tried to vary it a bit!  :smile: : 

I also replaced "outside" because that can feel a bit like "these people are outsiders" vs they just don't happen to be core. They are "inside" our community!

Same idea re: "side note"... maybe this is a Big Deal instead of a side note!

Author: This looks great to me! I couldn't find a clever way to write it, but inverting the statement like you did works awesomely
